File: src/main/php/g0201_0300/s0230_kth_smallest_element_in_a_bst/Solution.php
<?php

namespace leetcode\g0201_0300\s0230_kth_smallest_element_in_a_bst;

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Binary_Search_Tree #Data_Structure_II_Day_17_Tree #Level_2_Day_9_Binary_Search_Tree
// #Big_O_Time_O(n)_Space_O(n) #2023_12_23_Time_11_ms_(75.00%)_Space_22_MB_(96.88%)

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    private $k;
    private $count = 0;
    private $val;

    /**
     * @param TreeNode $root
     * @param Integer $k
     * @return Integer
     */
    public function kthSmallest($root, $k) {
        $this->k = $k;
        $this->calculate($root);
        return $this->val;
    }

    private function calculate($node) {
        if ($node->left == null && $node->right == null) {
            $this->count++;
            if ($this->count == $this->k) {
                $this->val = $node->val;
            }
            return;
        }
        if ($node->left != null) {
            $this->calculate($node->left);
        }
        $this->count++;
        if ($this->count == $this->k) {
            $this->val = $node->val;
            return;
        }
        if ($node->right != null) {
            $this->calculate($node->right);
        }
    }
}

File: src/main/python/g0201_0300/s0230_kth_smallest_element_in_a_bst/Solution.py
# #Medium #Top_100_Liked_Questions #Depth_First_Search #Tree #Binary_Tree #Binary_Search_Tree
# #Data_Structure_II_Day_17_Tree #Level_2_Day_9_Binary_Search_Tree #Big_O_Time_O(n)_Space_O(n)
# #2024_06_09_Time_44_ms_(61.91%)_Space_19.4_MB_(88.66%)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.k = 0
        self.count = 0
        self.val = None

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.k = k
        self.count = 0
        self.val = None
        self.calculate(root)
        return self.val

    def calculate(self, node):
        if not node:
            return

        if node.left:
            self.calculate(node.left)

        self.count += 1
        if self.count == self.k:
            self.val = node.val
            return

        if node.right:
            self.calculate(node.right)
